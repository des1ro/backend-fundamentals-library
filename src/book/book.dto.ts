export class Book {
  constructor(
    readonly title: string,
    readonly author: string,
    readonly year: number
  ) {}
}
