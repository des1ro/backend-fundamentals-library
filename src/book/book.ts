export class Book {
  constructor(
    readonly title: string,
    readonly author: string,
    readonly year: number
  ) {}
}
/**
 * book ->
 *  book.service.ts
 *  book.dto.ts
 *
 *
 * book ->
 *    services ->
 *      book.service.ts
 *      newspaper.service.ts
 *    dto ->
 *      book.dto.ts
 *       .....
 */
