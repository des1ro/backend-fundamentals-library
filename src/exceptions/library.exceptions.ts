type ErrorName = "PERMISION_DENIED" | "INVALID_BOOK" | "BOOKING_ERROR";
export class LibraryError extends Error {
  name: string;
  message: string;
  constructor({ name, message }: { name: ErrorName; message: string }) {
    super(name);
    this.name = name;
    this.message = message;
  }
}
