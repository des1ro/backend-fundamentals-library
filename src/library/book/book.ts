export class Book {
  constructor(
    private readonly title: string,
    private readonly author: string,
    private readonly year: number,
    private readonly id: number
  ) {}
  getId(): number {
    return this.id;
  }
  getTitle(): string {
    return this.title;
  }
  getAuthor(): string {
    return this.author;
  }
  getYear(): number {
    return this.year;
  }
}
