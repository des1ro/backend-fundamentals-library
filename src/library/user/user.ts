import { randomUUID } from "crypto";

export class User {
  private readonly uuid: string = randomUUID();
  private credit: number = 0;
  constructor(private readonly name: string) {}
  getName(): string {
    return this.name;
  }
  getUuid(): string {
    return this.uuid;
  }
  resetCredit() {
    this.credit = 0; // Czy tu powinienem wrzucić credit do construktora, (private readonly credit) i chcąc zmiany tworzyć new User?
  }
  getCredit() {
    return this.credit;
  }
  setCredit(credit: number): void {
    this.credit = credit;
  }
}
