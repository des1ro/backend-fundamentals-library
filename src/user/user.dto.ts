import { randomUUID } from "crypto";

export class User {
  readonly uuid: string = randomUUID();
  constructor(
    readonly name: string,
    private dueDate = 0,
    private penaltyPoints = 0
  ) {}
  getDueDate(): number {
    return this.dueDate;
  }
  getPenaltyPoints(): number {
    return this.penaltyPoints;
  }
  setDueDate(dueDate: number): void {
    this.dueDate = dueDate;
  }
  setPenaltyPoints(penaltyPoints: number): void {
    this.penaltyPoints = penaltyPoints;
  }
}
