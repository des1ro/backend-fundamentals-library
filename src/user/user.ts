import { randomUUID } from "crypto";
import { isAfter } from "date-fns";
export class User {
  readonly uuid: string = randomUUID();
  private dueDate?: number;
  private penaltyPoints = 0;
  constructor(readonly name: string) {}
  getPenaltyPoints() {
    return this.penaltyPoints;
  }
  addPenaltyPoints(penaltyPoints: number): void {
    this.penaltyPoints += penaltyPoints;
  }
  subtractPenaltyPoints(penaltyPoints: number): void {
    this.penaltyPoints -= penaltyPoints;
  }
  getDueDate() {
    return this.dueDate || 0;
  }
  setDueDate(date: number): void {
    this.dueDate = date;
  }
  isBanned(): boolean {
    return isAfter(this.getDueDate(), new Date());
  }
}
