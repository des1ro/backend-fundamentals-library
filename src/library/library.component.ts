import { Book } from "../book/book.dto";
import { LibraryError } from "../exceptions/library.exceptions";
import { User } from "../user/user.dto";
import { Booking } from "../booking/booking.service";
import { UsersUpdate } from "../user/service/usersUpdate/usersUpdate.service";
import * as cron from "node-cron";

export class Library {
  private readonly usersUpdate: UsersUpdate;
  constructor(
    private readonly libraryBooks = new Map<Book, number>(),
    private readonly bookings = new Map<User, Booking>()
  ) {
    this.usersUpdate = new UsersUpdate();
  }
  constantlyUpdateUsers(): void {
    cron.schedule("* 30 5 * * * ", () => {
      console.log("Update users every day at 5:30");
      this.usersUpdate.update(this.bookings);
      console.log("Library users updated");
    });
  }
  addBook(book: Book, amount = 1): void {
    const value = this.libraryBooks.get(book) || 0;
    this.libraryBooks.set(book, amount + value);
  }
  deleteBook(book: Book, amount: number = 1): void {
    const currentAmount = this.libraryBooks.get(book);
    if (!currentAmount) {
      throw new LibraryError({
        name: "INVALID_BOOK",
        message: "The book is not available in this library",
      });
    }
    const newAmount = currentAmount - amount;
    if (newAmount > 0) {
      this.libraryBooks.set(book, newAmount);
      return;
    }
    this.libraryBooks.delete(book);
  }
  createBooking(user: User) {
    if (this.bookings.has(user)) {
      throw new LibraryError({
        name: "BOOKING_ERROR",
        message: "Booking already exist",
      });
    }
    this.bookings.set(user, new Booking(user));
  }
  borrowBook(user: User, book: Book) {
    const booking = this.bookings.get(user);
    const currentAmount = this.libraryBooks.get(book) || 0;
    if (currentAmount === 0) {
      throw new LibraryError({
        name: "INVALID_BOOK",
        message: "Book isn't in library",
      });
    }
    if (booking) {
      this.libraryBooks.set(book, currentAmount - 1);
      booking.addBookToUser(book);
      return;
    }
    throw new LibraryError({
      name: "BOOKING_ERROR",
      message: "Booking doen't exist",
    });
  }
  returnBook(user: User, book: Book) {
    const booking = this.bookings.get(user);
    if (!booking) {
      throw new LibraryError({
        name: "INVALID_BOOK",
        message: "Book not found",
      });
    }
    const currentAmount = this.libraryBooks.get(book);
    if (currentAmount !== undefined) {
      booking.returnBook(book);
      this.libraryBooks.set(book, currentAmount + 1);
      return;
    }
    throw new LibraryError({
      name: "INVALID_BOOK",
      message: "Book not found in library",
    });
  }
}
