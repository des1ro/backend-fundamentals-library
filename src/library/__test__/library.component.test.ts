import { Book } from "../../book/book.dto";
import { LibraryError } from "../../exceptions/library.exceptions";
import { User } from "../../user/user.dto";
import { Library } from "../library.component";
jest.mock("crypto", () => ({
  randomUUID: jest.fn().mockReturnValue("test-uuid"),
}));
jest.mock("node-cron", () => {
  return {
    schedule: jest.fn(),
  };
});
describe("Library test suite", () => {
  let objectUnderTest: Library;
  let mockedUserOne: User;
  let mockedBookOne: Book;
  let mockedBookTwo: Book;
  beforeAll(() => {
    mockedBookOne = new Book("Test title One", "Test author One", 1992);
    mockedBookTwo = new Book("Test title Two", "Test author Two", 1995);
  });
  beforeEach(() => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date(2023, 0, 20, 12, 30, 0, 0));
    mockedUserOne = new User("Test user one");
    objectUnderTest = new Library();
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  afterAll(() => {
    jest.restoreAllMocks();
  });

  it("Should add book to library properly", () => {
    //When
    const resultOne = objectUnderTest;
    objectUnderTest.addBook(mockedBookOne);
    const resultTwo = objectUnderTest;
    //Then
    expect(resultOne).toMatchSnapshot();
    expect(resultTwo).toMatchSnapshot();
  });
  describe("Delete book test suite", () => {
    it("Should set books amoount properly", () => {
      //Given
      jest.spyOn(objectUnderTest, "deleteBook");
      //When
      objectUnderTest.addBook(mockedBookOne, 10);
      const resultOne = objectUnderTest;
      objectUnderTest.deleteBook(mockedBookOne, 2);
      const resultTwo = objectUnderTest;
      //Then
      expect(resultOne).toMatchSnapshot();
      expect(resultTwo).toMatchSnapshot();
      expect(objectUnderTest.deleteBook).toBeCalledTimes(1);
    });
    it("Should delete book if if delete more than is in library", () => {
      //Given
      jest.spyOn(objectUnderTest, "deleteBook");
      //When
      objectUnderTest.addBook(mockedBookOne, 10);
      const resultOne = objectUnderTest;
      objectUnderTest.deleteBook(mockedBookOne, 11);
      const resultTwo = objectUnderTest;
      //Then
      expect(resultOne).toMatchSnapshot();
      expect(resultTwo).toMatchSnapshot();
      expect(objectUnderTest.deleteBook).toBeCalledTimes(1);
    });
    it("Should throw an LibraryError if book isn't in library", () => {
      //When
      const result = objectUnderTest;
      //Then
      expect(result).toMatchSnapshot();
      expect(() => objectUnderTest.deleteBook(mockedBookOne)).toThrow(
        LibraryError
      );
    });
  });
  describe("Create booking test suite", () => {
    it("Should create booking correctly", () => {
      //Given
      jest.spyOn(objectUnderTest, "createBooking");
      //When
      const resultOne = objectUnderTest;
      objectUnderTest.createBooking(mockedUserOne);
      const resultTwo = objectUnderTest;
      //Then
      expect(resultOne).toMatchSnapshot();
      expect(resultTwo).toMatchSnapshot();
      expect(objectUnderTest.createBooking).toBeCalledTimes(1);
    });
    it("Should throw library error if booking with this user already exist", () => {
      //Given
      jest.spyOn(objectUnderTest, "createBooking");
      //When
      objectUnderTest.createBooking(mockedUserOne);
      const result = objectUnderTest;
      //Then
      expect(result).toMatchSnapshot();
      expect(() => objectUnderTest.createBooking(mockedUserOne)).toThrow(
        LibraryError
      );
      expect(objectUnderTest.createBooking).toBeCalledTimes(2);
    });
  });

  describe("Borrow book suite test", () => {
    it("Should set new amount of books if user borrow", () => {
      //Given
      objectUnderTest.createBooking(mockedUserOne);
      objectUnderTest.addBook(mockedBookOne);
      const resultOne = objectUnderTest;
      //When
      objectUnderTest.borrowBook(mockedUserOne, mockedBookOne);
      const resultTwo = objectUnderTest;
      //Then
      expect(resultOne).toMatchSnapshot();
      expect(resultTwo).toMatchSnapshot();
    });
    it("Should throw an LibraryError if book isn't in library", () => {
      //Given
      objectUnderTest.createBooking(mockedUserOne);
      objectUnderTest.addBook(mockedBookOne);
      //When
      const resultOne = objectUnderTest;
      jest.spyOn(objectUnderTest, "borrowBook");
      //Then
      expect(() =>
        objectUnderTest.borrowBook(mockedUserOne, mockedBookTwo)
      ).toThrow(LibraryError);
      expect(objectUnderTest.borrowBook).toBeCalledTimes(1);
      expect(resultOne).toMatchSnapshot();
    });
    it("Should throw an LibraryError if booking don't exist", () => {
      //Given
      objectUnderTest.addBook(mockedBookOne);
      jest.spyOn(objectUnderTest, "borrowBook");
      //When
      const result = objectUnderTest;
      //Then
      expect(() =>
        objectUnderTest.borrowBook(mockedUserOne, mockedBookOne)
      ).toThrow(LibraryError);
      expect(result).toMatchSnapshot();
      expect(objectUnderTest.borrowBook).toBeCalledTimes(1);
    });
  });
  describe("Return book suite test", () => {
    it("Should return book properly and set new value in library", () => {
      //Given
      jest.spyOn(objectUnderTest, "returnBook");
      objectUnderTest.createBooking(mockedUserOne);
      objectUnderTest.addBook(mockedBookOne, 5);
      const resultOne = objectUnderTest;
      //When
      objectUnderTest.borrowBook(mockedUserOne, mockedBookOne);
      const resultTwo = objectUnderTest;
      objectUnderTest.returnBook(mockedUserOne, mockedBookOne);
      //Then
      expect(resultOne).toMatchSnapshot();
      expect(resultTwo).toMatchSnapshot();
      expect(objectUnderTest.returnBook).toBeCalledTimes(1);
    });
    it("Should throw an LibraryError on return wrong book", () => {
      //Given
      objectUnderTest.createBooking(mockedUserOne);
      objectUnderTest.addBook(mockedBookOne);
      //When
      const resultOne = objectUnderTest;
      objectUnderTest.borrowBook(mockedUserOne, mockedBookOne);
      const resultTwo = objectUnderTest;
      //Then
      expect(resultOne).toMatchSnapshot();
      expect(resultTwo).toMatchSnapshot();
      expect(() =>
        objectUnderTest.returnBook(mockedUserOne, mockedBookTwo)
      ).toThrow(LibraryError);
    });
    it("Should throw an LibraryError if booking not exist", () => {
      //Given
      objectUnderTest.addBook(mockedBookOne);
      //When
      const resultOne = objectUnderTest;
      //Then
      expect(resultOne).toMatchSnapshot();
      expect(() =>
        objectUnderTest.returnBook(mockedUserOne, mockedBookTwo)
      ).toThrow(LibraryError);
    });
  });
});
