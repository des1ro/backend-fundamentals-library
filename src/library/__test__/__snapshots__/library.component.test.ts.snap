// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Library test suite Borrow book suite test Should set new amount of books if user borrow 1`] = `
Library {
  "bookings": Map {
    User {
      "dueDate": 0,
      "name": "Test user one",
      "penaltyPoints": 0,
      "uuid": "test-uuid",
    } => Booking {
      "booksBorrowDate": Map {
        Book {
          "author": "Test author One",
          "title": "Test title One",
          "year": 1992,
        } => 2023-01-20T11:30:00.000Z,
      },
      "user": User {
        "dueDate": 0,
        "name": "Test user one",
        "penaltyPoints": 0,
        "uuid": "test-uuid",
      },
    },
  },
  "libraryBooks": Map {
    Book {
      "author": "Test author One",
      "title": "Test title One",
      "year": 1992,
    } => 0,
  },
  "usersUpdate": UsersUpdate {
    "penaltyPointsCalc": PenaltyPoints {
      "borrowMonthsLimit": 1,
      "penaltyPointsPerBook": 10,
    },
  },
}
`;

exports[`Library test suite Borrow book suite test Should set new amount of books if user borrow 2`] = `
Library {
  "bookings": Map {
    User {
      "dueDate": 0,
      "name": "Test user one",
      "penaltyPoints": 0,
      "uuid": "test-uuid",
    } => Booking {
      "booksBorrowDate": Map {
        Book {
          "author": "Test author One",
          "title": "Test title One",
          "year": 1992,
        } => 2023-01-20T11:30:00.000Z,
      },
      "user": User {
        "dueDate": 0,
        "name": "Test user one",
        "penaltyPoints": 0,
        "uuid": "test-uuid",
      },
    },
  },
  "libraryBooks": Map {
    Book {
      "author": "Test author One",
      "title": "Test title One",
      "year": 1992,
    } => 0,
  },
  "usersUpdate": UsersUpdate {
    "penaltyPointsCalc": PenaltyPoints {
      "borrowMonthsLimit": 1,
      "penaltyPointsPerBook": 10,
    },
  },
}
`;

exports[`Library test suite Borrow book suite test Should throw an LibraryError if book isn't in library 1`] = `
Library {
  "bookings": Map {
    User {
      "dueDate": 0,
      "name": "Test user one",
      "penaltyPoints": 0,
      "uuid": "test-uuid",
    } => Booking {
      "booksBorrowDate": Map {},
      "user": User {
        "dueDate": 0,
        "name": "Test user one",
        "penaltyPoints": 0,
        "uuid": "test-uuid",
      },
    },
  },
  "borrowBook": [MockFunction] {
    "calls": [
      [
        User {
          "dueDate": 0,
          "name": "Test user one",
          "penaltyPoints": 0,
          "uuid": "test-uuid",
        },
        Book {
          "author": "Test author Two",
          "title": "Test title Two",
          "year": 1995,
        },
      ],
    ],
    "results": [
      {
        "type": "throw",
        "value": [INVALID_BOOK: Book isn't in library],
      },
    ],
  },
  "libraryBooks": Map {
    Book {
      "author": "Test author One",
      "title": "Test title One",
      "year": 1992,
    } => 1,
  },
  "usersUpdate": UsersUpdate {
    "penaltyPointsCalc": PenaltyPoints {
      "borrowMonthsLimit": 1,
      "penaltyPointsPerBook": 10,
    },
  },
}
`;

exports[`Library test suite Borrow book suite test Should throw an LibraryError if booking don't exist 1`] = `
Library {
  "bookings": Map {},
  "borrowBook": [MockFunction] {
    "calls": [
      [
        User {
          "dueDate": 0,
          "name": "Test user one",
          "penaltyPoints": 0,
          "uuid": "test-uuid",
        },
        Book {
          "author": "Test author One",
          "title": "Test title One",
          "year": 1992,
        },
      ],
    ],
    "results": [
      {
        "type": "throw",
        "value": [BOOKING_ERROR: Booking doen't exist],
      },
    ],
  },
  "libraryBooks": Map {
    Book {
      "author": "Test author One",
      "title": "Test title One",
      "year": 1992,
    } => 1,
  },
  "usersUpdate": UsersUpdate {
    "penaltyPointsCalc": PenaltyPoints {
      "borrowMonthsLimit": 1,
      "penaltyPointsPerBook": 10,
    },
  },
}
`;

exports[`Library test suite Create booking test suite Should create booking correctly 1`] = `
Library {
  "bookings": Map {
    User {
      "dueDate": 0,
      "name": "Test user one",
      "penaltyPoints": 0,
      "uuid": "test-uuid",
    } => Booking {
      "booksBorrowDate": Map {},
      "user": User {
        "dueDate": 0,
        "name": "Test user one",
        "penaltyPoints": 0,
        "uuid": "test-uuid",
      },
    },
  },
  "createBooking": [MockFunction] {
    "calls": [
      [
        User {
          "dueDate": 0,
          "name": "Test user one",
          "penaltyPoints": 0,
          "uuid": "test-uuid",
        },
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "libraryBooks": Map {},
  "usersUpdate": UsersUpdate {
    "penaltyPointsCalc": PenaltyPoints {
      "borrowMonthsLimit": 1,
      "penaltyPointsPerBook": 10,
    },
  },
}
`;

exports[`Library test suite Create booking test suite Should create booking correctly 2`] = `
Library {
  "bookings": Map {
    User {
      "dueDate": 0,
      "name": "Test user one",
      "penaltyPoints": 0,
      "uuid": "test-uuid",
    } => Booking {
      "booksBorrowDate": Map {},
      "user": User {
        "dueDate": 0,
        "name": "Test user one",
        "penaltyPoints": 0,
        "uuid": "test-uuid",
      },
    },
  },
  "createBooking": [MockFunction] {
    "calls": [
      [
        User {
          "dueDate": 0,
          "name": "Test user one",
          "penaltyPoints": 0,
          "uuid": "test-uuid",
        },
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "libraryBooks": Map {},
  "usersUpdate": UsersUpdate {
    "penaltyPointsCalc": PenaltyPoints {
      "borrowMonthsLimit": 1,
      "penaltyPointsPerBook": 10,
    },
  },
}
`;

exports[`Library test suite Create booking test suite Should throw library error if booking with this user already exist 1`] = `
Library {
  "bookings": Map {
    User {
      "dueDate": 0,
      "name": "Test user one",
      "penaltyPoints": 0,
      "uuid": "test-uuid",
    } => Booking {
      "booksBorrowDate": Map {},
      "user": User {
        "dueDate": 0,
        "name": "Test user one",
        "penaltyPoints": 0,
        "uuid": "test-uuid",
      },
    },
  },
  "createBooking": [MockFunction] {
    "calls": [
      [
        User {
          "dueDate": 0,
          "name": "Test user one",
          "penaltyPoints": 0,
          "uuid": "test-uuid",
        },
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "libraryBooks": Map {},
  "usersUpdate": UsersUpdate {
    "penaltyPointsCalc": PenaltyPoints {
      "borrowMonthsLimit": 1,
      "penaltyPointsPerBook": 10,
    },
  },
}
`;

exports[`Library test suite Delete book test suite Should delete book if if delete more than is in library 1`] = `
Library {
  "bookings": Map {},
  "deleteBook": [MockFunction] {
    "calls": [
      [
        Book {
          "author": "Test author One",
          "title": "Test title One",
          "year": 1992,
        },
        11,
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "libraryBooks": Map {},
  "usersUpdate": UsersUpdate {
    "penaltyPointsCalc": PenaltyPoints {
      "borrowMonthsLimit": 1,
      "penaltyPointsPerBook": 10,
    },
  },
}
`;

exports[`Library test suite Delete book test suite Should delete book if if delete more than is in library 2`] = `
Library {
  "bookings": Map {},
  "deleteBook": [MockFunction] {
    "calls": [
      [
        Book {
          "author": "Test author One",
          "title": "Test title One",
          "year": 1992,
        },
        11,
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "libraryBooks": Map {},
  "usersUpdate": UsersUpdate {
    "penaltyPointsCalc": PenaltyPoints {
      "borrowMonthsLimit": 1,
      "penaltyPointsPerBook": 10,
    },
  },
}
`;

exports[`Library test suite Delete book test suite Should set books amoount properly 1`] = `
Library {
  "bookings": Map {},
  "deleteBook": [MockFunction] {
    "calls": [
      [
        Book {
          "author": "Test author One",
          "title": "Test title One",
          "year": 1992,
        },
        2,
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "libraryBooks": Map {
    Book {
      "author": "Test author One",
      "title": "Test title One",
      "year": 1992,
    } => 8,
  },
  "usersUpdate": UsersUpdate {
    "penaltyPointsCalc": PenaltyPoints {
      "borrowMonthsLimit": 1,
      "penaltyPointsPerBook": 10,
    },
  },
}
`;

exports[`Library test suite Delete book test suite Should set books amoount properly 2`] = `
Library {
  "bookings": Map {},
  "deleteBook": [MockFunction] {
    "calls": [
      [
        Book {
          "author": "Test author One",
          "title": "Test title One",
          "year": 1992,
        },
        2,
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "libraryBooks": Map {
    Book {
      "author": "Test author One",
      "title": "Test title One",
      "year": 1992,
    } => 8,
  },
  "usersUpdate": UsersUpdate {
    "penaltyPointsCalc": PenaltyPoints {
      "borrowMonthsLimit": 1,
      "penaltyPointsPerBook": 10,
    },
  },
}
`;

exports[`Library test suite Delete book test suite Should throw an LibraryError if book isn't in library 1`] = `
Library {
  "bookings": Map {},
  "libraryBooks": Map {},
  "usersUpdate": UsersUpdate {
    "penaltyPointsCalc": PenaltyPoints {
      "borrowMonthsLimit": 1,
      "penaltyPointsPerBook": 10,
    },
  },
}
`;

exports[`Library test suite Return book suite test Should return book properly and set new value in library 1`] = `
Library {
  "bookings": Map {
    User {
      "dueDate": 0,
      "name": "Test user one",
      "penaltyPoints": 0,
      "uuid": "test-uuid",
    } => Booking {
      "booksBorrowDate": Map {},
      "user": User {
        "dueDate": 0,
        "name": "Test user one",
        "penaltyPoints": 0,
        "uuid": "test-uuid",
      },
    },
  },
  "libraryBooks": Map {
    Book {
      "author": "Test author One",
      "title": "Test title One",
      "year": 1992,
    } => 5,
  },
  "returnBook": [MockFunction] {
    "calls": [
      [
        User {
          "dueDate": 0,
          "name": "Test user one",
          "penaltyPoints": 0,
          "uuid": "test-uuid",
        },
        Book {
          "author": "Test author One",
          "title": "Test title One",
          "year": 1992,
        },
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "usersUpdate": UsersUpdate {
    "penaltyPointsCalc": PenaltyPoints {
      "borrowMonthsLimit": 1,
      "penaltyPointsPerBook": 10,
    },
  },
}
`;

exports[`Library test suite Return book suite test Should return book properly and set new value in library 2`] = `
Library {
  "bookings": Map {
    User {
      "dueDate": 0,
      "name": "Test user one",
      "penaltyPoints": 0,
      "uuid": "test-uuid",
    } => Booking {
      "booksBorrowDate": Map {},
      "user": User {
        "dueDate": 0,
        "name": "Test user one",
        "penaltyPoints": 0,
        "uuid": "test-uuid",
      },
    },
  },
  "libraryBooks": Map {
    Book {
      "author": "Test author One",
      "title": "Test title One",
      "year": 1992,
    } => 5,
  },
  "returnBook": [MockFunction] {
    "calls": [
      [
        User {
          "dueDate": 0,
          "name": "Test user one",
          "penaltyPoints": 0,
          "uuid": "test-uuid",
        },
        Book {
          "author": "Test author One",
          "title": "Test title One",
          "year": 1992,
        },
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "usersUpdate": UsersUpdate {
    "penaltyPointsCalc": PenaltyPoints {
      "borrowMonthsLimit": 1,
      "penaltyPointsPerBook": 10,
    },
  },
}
`;

exports[`Library test suite Return book suite test Should throw an LibraryError if booking not exist 1`] = `
Library {
  "bookings": Map {},
  "libraryBooks": Map {
    Book {
      "author": "Test author One",
      "title": "Test title One",
      "year": 1992,
    } => 1,
  },
  "usersUpdate": UsersUpdate {
    "penaltyPointsCalc": PenaltyPoints {
      "borrowMonthsLimit": 1,
      "penaltyPointsPerBook": 10,
    },
  },
}
`;

exports[`Library test suite Return book suite test Should throw an LibraryError on return wrong book 1`] = `
Library {
  "bookings": Map {
    User {
      "dueDate": 0,
      "name": "Test user one",
      "penaltyPoints": 0,
      "uuid": "test-uuid",
    } => Booking {
      "booksBorrowDate": Map {
        Book {
          "author": "Test author One",
          "title": "Test title One",
          "year": 1992,
        } => 2023-01-20T11:30:00.000Z,
      },
      "user": User {
        "dueDate": 0,
        "name": "Test user one",
        "penaltyPoints": 0,
        "uuid": "test-uuid",
      },
    },
  },
  "libraryBooks": Map {
    Book {
      "author": "Test author One",
      "title": "Test title One",
      "year": 1992,
    } => 0,
  },
  "usersUpdate": UsersUpdate {
    "penaltyPointsCalc": PenaltyPoints {
      "borrowMonthsLimit": 1,
      "penaltyPointsPerBook": 10,
    },
  },
}
`;

exports[`Library test suite Return book suite test Should throw an LibraryError on return wrong book 2`] = `
Library {
  "bookings": Map {
    User {
      "dueDate": 0,
      "name": "Test user one",
      "penaltyPoints": 0,
      "uuid": "test-uuid",
    } => Booking {
      "booksBorrowDate": Map {
        Book {
          "author": "Test author One",
          "title": "Test title One",
          "year": 1992,
        } => 2023-01-20T11:30:00.000Z,
      },
      "user": User {
        "dueDate": 0,
        "name": "Test user one",
        "penaltyPoints": 0,
        "uuid": "test-uuid",
      },
    },
  },
  "libraryBooks": Map {
    Book {
      "author": "Test author One",
      "title": "Test title One",
      "year": 1992,
    } => 0,
  },
  "usersUpdate": UsersUpdate {
    "penaltyPointsCalc": PenaltyPoints {
      "borrowMonthsLimit": 1,
      "penaltyPointsPerBook": 10,
    },
  },
}
`;

exports[`Library test suite Should add book to library properly 1`] = `
Library {
  "bookings": Map {},
  "libraryBooks": Map {
    Book {
      "author": "Test author One",
      "title": "Test title One",
      "year": 1992,
    } => 1,
  },
  "usersUpdate": UsersUpdate {
    "penaltyPointsCalc": PenaltyPoints {
      "borrowMonthsLimit": 1,
      "penaltyPointsPerBook": 10,
    },
  },
}
`;

exports[`Library test suite Should add book to library properly 2`] = `
Library {
  "bookings": Map {},
  "libraryBooks": Map {
    Book {
      "author": "Test author One",
      "title": "Test title One",
      "year": 1992,
    } => 1,
  },
  "usersUpdate": UsersUpdate {
    "penaltyPointsCalc": PenaltyPoints {
      "borrowMonthsLimit": 1,
      "penaltyPointsPerBook": 10,
    },
  },
}
`;
